### PR TITLE
AUGRUCell fusion transformation

### PR DESCRIPTION
**Pull request by itikhono**
### Details:
 *This transformation detects pattern of AUGRU Cell operation in the graph and fuse it to our internal operation*

### Tickets:
 - *76704*

### CODE REVIEW

File: src/tests/functional/inference_engine/transformations/common_optimizations/augru_cell_fusion.cpp [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+
+class AUGRUFusionTestDyn
+        : public testing::WithParamInterface<AUGRUFusionParams>, public TransformationTestsF {
+};
+
+TEST_P(AUGRUFusionTestDyn, AUGRUCellPatternDynamicShapes) {
+    auto p = GetParam();
+    {
+        function = gen_model(p.batch, p.hidden_size, p.input_size, true);
+        manager.register_pass<ov::pass::AUGRUCellFusion>(); // the transformation won't be applied
```
[the remark this answers is not in the document]
Comment (itikhono) [reply src/tests/functional/inference_engine/transformations/common_optimizations/augru_cell_fusion.cpp:116]: 1. this type of tests has some internal logic, if we don't create reference model in the test, by default, a clone of the original model will be used as reference. So I think it's enough to compare nodes in the model only. Node comparison is enabled by default.
2. we need to know hidden_size and input_size, these dimensions must be static
image
batch, seq_lengths can be dynamic

File: src/tests/functional/inference_engine/transformations/common_optimizations/augru_cell_fusion.cpp [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+}
+
+class AUGRUFusionTestDyn
+        : public testing::WithParamInterface<AUGRUFusionParams>, public TransformationTestsF {
+};
+
+TEST_P(AUGRUFusionTestDyn, AUGRUCellPatternDynamicShapes) {
+    auto p = GetParam();
+    {
+        function = gen_model(p.batch, p.hidden_size, p.input_size, true);
```
[the remark this answers is not in the document]
Comment (itikhono) [reply src/tests/functional/inference_engine/transformations/common_optimizations/augru_cell_fusion.cpp:115]: Yes, I think so

File: src/common/transformations/src/transformations/common_optimizations/augru_cell_fusion.cpp [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+        auto split_lenghts = make_shared<Constant>(element::i64, Shape{2}, vector<int64_t>{input_size, hidden_size});
+        auto split_WRzr = make_shared<VariadicSplit>(WRzr, axis_1, split_lenghts);
+        auto split_WRh = make_shared<VariadicSplit>(WRh, axis_1, split_lenghts);
+        auto Wzrh = make_shared<Concat>(ov::OutputVector{split_WRzr->output(0), split_WRh->output(0)}, 0);
+        auto Rzrh = make_shared<Concat>(ov::OutputVector{split_WRzr->output(1), split_WRh->output(1)}, 0);
+
+        auto squeeze_B = make_shared<Squeeze>(B, axis_0);
+        auto cell = make_shared<ov::op::internal::AUGRUCell>(X, H, Wzrh, Rzrh, squeeze_B, A, H.get_shape()[1]);
+
+        NodeVector new_nodes;
```
[the remark this answers is not in the document]
Comment (itikhono) [reply src/common/transformations/src/transformations/common_optimizations/augru_cell_fusion.cpp:82]: Sure

File: src/tests/functional/inference_engine/transformations/common_optimizations/augru_cell_fusion.cpp [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+    size_t input_size;
+};
+
+class AUGRUFusionTest
+        : public testing::WithParamInterface<AUGRUFusionParams>,
+          public TransformationTestsF {
+};
+
+TEST_P(AUGRUFusionTest, AUGRUCellPattern) {
+    auto p = GetParam();
```
[the remark this answers is not in the document]
Comment (itikhono) [reply src/tests/functional/inference_engine/transformations/common_optimizations/augru_cell_fusion.cpp]: fixed

File: src/tests/functional/inference_engine/transformations/common_optimizations/augru_cell_fusion.cpp [hunk 5]
```diff
@@ -0,0 +1,126 @@
+// Copyright (C) 2018-2022 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include <gtest/gtest.h>
```
[the remark this answers is not in the document]
Comment (itikhono) [reply src/tests/functional/inference_engine/transformations/common_optimizations/augru_cell_fusion.cpp]: done

File: src/common/transformations/src/transformations/common_optimizations/augru_cell_fusion.cpp [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+        auto axis_1 = make_shared<Constant>(element::i64, Shape{}, 1);
+        auto B = make_shared<Concat>(ov::OutputVector{bias_add_1->input_value(1), bias_add_2->input_value(1)}, 1);
+
+        auto WRzr = pattern_map.at(matmul_1)->input_value(1);
+        auto WRh = pattern_map.at(matmul_2)->input_value(1);
+
+        auto split_lenghts = make_shared<Constant>(element::i64, Shape{2}, vector<int64_t>{input_size, hidden_size});
+        auto split_WRzr = make_shared<VariadicSplit>(WRzr, axis_1, split_lenghts);
+        auto split_WRh = make_shared<VariadicSplit>(WRh, axis_1, split_lenghts);
+        auto Wzrh = make_shared<Concat>(ov::OutputVector{split_WRzr->output(0), split_WRh->output(0)}, 0);
```
[the remark this answers is not in the document]
Comment (itikhono) [reply src/common/transformations/src/transformations/common_optimizations/augru_cell_fusion.cpp:76]: done

File: src/common/transformations/include/transformations/common_optimizations/augru_cell_fusion.hpp [hunk 7]
```diff
@@ -0,0 +1,35 @@
+// Copyright (C) 2018-2022 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#pragma once
+
+#include <memory>
+#include <openvino/pass/graph_rewrite.hpp>
```
[the remark this answers is not in the document]
Comment (itikhono) [reply src/common/transformations/include/transformations/common_optimizations/augru_cell_fusion.hpp:8]: done

File: src/tests/functional/inference_engine/transformations/common_optimizations/augru_cell_fusion.cpp [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+    auto split_WRzr = make_shared<VariadicSplit>(WRzr, axis_1, split_lenghts);
+    auto split_WRh = make_shared<VariadicSplit>(WRh, axis_1, split_lenghts);
+    auto Wzrh = make_shared<Concat>(ov::OutputVector{split_WRzr->output(0), split_WRh->output(0)}, 0);
+    auto Rzrh = make_shared<Concat>(ov::OutputVector{split_WRzr->output(1), split_WRh->output(1)}, 0);
+
+    auto B = make_shared<Concat>(ov::OutputVector{Bzr, Bh}, 1);
+
+    auto squeeze_B = make_shared<Squeeze>(B, axis_0);
+    auto cell = make_shared<ov::op::internal::AUGRUCell>(X, H, Wzrh, Rzrh, squeeze_B, A, hidden_size);
+    return std::make_shared<ngraph::Function>(ov::OutputVector {cell}, params);
```
[the remark this answers is not in the document]
Comment (itikhono) [reply src/tests/functional/inference_engine/transformations/common_optimizations/augru_cell_fusion.cpp]: fixed

File: src/tests/functional/inference_engine/transformations/common_optimizations/augru_cell_fusion.cpp [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+    }
+}
+
+static std::vector<AUGRUFusionParams> params = {
+        AUGRUFusionParams{1, 1, 1},
+        AUGRUFusionParams{2, 128, 32},
+};
+
+INSTANTIATE_TEST_SUITE_P(AUGRUFusionTest, AUGRUFusionTest, ::testing::ValuesIn(params));
+INSTANTIATE_TEST_SUITE_P(AUGRUFusionTestDyn, AUGRUFusionTestDyn, ::testing::ValuesIn(params));
```
[the remark this answers is not in the document]
Comment (itikhono) [reply src/tests/functional/inference_engine/transformations/common_optimizations/augru_cell_fusion.cpp]: done

File: src/tests/functional/inference_engine/transformations/common_optimizations/augru_cell_fusion.cpp [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+
+TEST_P(AUGRUFusionTestDyn, AUGRUCellPatternDynamicShapes) {
+    auto p = GetParam();
+    {
+        function = gen_model(p.batch, p.hidden_size, p.input_size, true);
+        manager.register_pass<ov::pass::AUGRUCellFusion>(); // the transformation won't be applied
+    }
+}
+
+static std::vector<AUGRUFusionParams> params = {
```
[the remark this answers is not in the document]
Comment (itikhono) [reply src/tests/functional/inference_engine/transformations/common_optimizations/augru_cell_fusion.cpp]: done

File: src/tests/functional/inference_engine/transformations/common_optimizations/augru_cell_fusion.cpp [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+    comparator.enable(FunctionsComparator::CmpValues::CONST_VALUES);
+    comparator.enable(FunctionsComparator::CmpValues::ATTRIBUTES);
+}
+
+class AUGRUFusionTestDyn
+        : public testing::WithParamInterface<AUGRUFusionParams>, public TransformationTestsF {
+};
+
+TEST_P(AUGRUFusionTestDyn, AUGRUCellPatternDynamicShapes) {
+    auto p = GetParam();
```
[the remark this answers is not in the document]
Comment (itikhono) [reply src/tests/functional/inference_engine/transformations/common_optimizations/augru_cell_fusion.cpp]: fixed

File: src/tests/functional/inference_engine/transformations/common_optimizations/augru_cell_fusion.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (itikhono) [reply src/tests/functional/inference_engine/transformations/common_optimizations/augru_cell_fusion.cpp]: done

File: src/common/transformations/src/transformations/common_optimizations/augru_cell_fusion.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (itikhono) [reply src/common/transformations/src/transformations/common_optimizations/augru_cell_fusion.cpp:86]: done

**Review by praasz — APPROVED** [review]

File: src/common/transformations/include/transformations/common_optimizations/augru_cell_fusion.hpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (ilya-lavrenov) [reply src/common/transformations/include/transformations/common_optimizations/augru_cell_fusion.hpp:8]: > Detail: Use `""` for project includes, use same style as in `cpp` file.

BTW, do you how we can tune static code checkers to auto-detect this?

[the remark this answers is not in the document]
Comment (praasz) [reply src/common/transformations/include/transformations/common_optimizations/augru_cell_fusion.hpp:8]: @ilya-lavrenov That would be great to have tool for it but I don't know which can do that.

[the remark this answers is not in the document]
Comment (ilya-lavrenov) [reply src/common/transformations/include/transformations/common_optimizations/augru_cell_fusion.hpp:8]: we have cpplint and clang_format.. But not sure whether they can be configurable to complain on such cases

[the remark this answers is not in the document]
Comment (praasz) [reply src/common/transformations/include/transformations/common_optimizations/augru_cell_fusion.hpp:8]: Yes looked at these tools and not found good option to catch it.

[the remark this answers is not in the document]
Comment (ilyachur) [reply src/common/transformations/include/transformations/common_optimizations/augru_cell_fusion.hpp:8]: The comment wasn't solved

**Review by ilyachur — COMMENTED** [review]
I don't think that you need to update onednn submodule in this PR

File: src/common/transformations/src/transformations/common_optimizations/augru_cell_fusion.cpp [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+ov::pass::AUGRUCellFusion::AUGRUCellFusion() {
+    MATCHER_SCOPE(AUGRUCellFusion);
+
+    // we can't determine hidden_size or input_size in this case
+    const auto is_first_dim_dynamic = [](const Output<Node>& output) -> bool {
+        const auto& p_shape = output.get_partial_shape();
+        return !(p_shape.rank().is_dynamic() || p_shape[1].is_dynamic());
+    };
+
+    auto concat_1 = wrap_type<Concat>({any_input(is_first_dim_dynamic), any_input(is_first_dim_dynamic)});
```
Comment (mitruska) [inline src/common/transformations/src/transformations/common_optimizations/augru_cell_fusion.cpp:30]: Shouldn't be `is_second_dim_static` check instead? ;)

**Review by mitruska — COMMENTED** [review]
In general LGTM, but I can see a potential issue, in case of (update, reset) gates layout instead of (reset. update) for Weights/Bias inputs.

(_The same comments for GRUCell fusion PR https://github.com/openvinotoolkit/openvino/pull/12843)_

File: src/common/transformations/include/transformations/common_optimizations/augru_cell_fusion.hpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (itikhono) [reply src/common/transformations/include/transformations/common_optimizations/augru_cell_fusion.hpp]: fixed

File: src/common/transformations/src/transformations/common_optimizations/augru_cell_fusion.cpp [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+        auto A = pattern_map.at(subtract_1)->input_value(1);
+        // biases are required
+        auto bias_add_1 = pattern_map.at(add_1);
+        auto bias_add_2 = pattern_map.at(add_2);
+
+        auto axis_0 = rg.make<Constant>(i64, Shape{}, 0);
+        auto axis_1 = rg.make<Constant>(i64, Shape{}, 1);
+        auto B = rg.make<Concat>(OutputVector{bias_add_1->input_value(1), bias_add_2->input_value(1)}, 1);
+
+        auto WRzr = pattern_map.at(matmul_1)->input_value(1);
```
[the remark this answers is not in the document]
Comment (itikhono) [reply src/common/transformations/src/transformations/common_optimizations/augru_cell_fusion.cpp]: supported rzh format
As a result of the discussion, we decided to support rzh only

File: src/common/transformations/src/transformations/common_optimizations/augru_cell_fusion.cpp [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+        auto input_size = x_pshape[1].get_length();
+
+        auto A = pattern_map.at(subtract_1)->input_value(1);
+        // biases are required
+        auto bias_add_1 = pattern_map.at(add_1);
+        auto bias_add_2 = pattern_map.at(add_2);
+
+        auto axis_0 = rg.make<Constant>(i64, Shape{}, 0);
+        auto axis_1 = rg.make<Constant>(i64, Shape{}, 1);
+        auto B = rg.make<Concat>(OutputVector{bias_add_1->input_value(1), bias_add_2->input_value(1)}, 1);
```
[the remark this answers is not in the document]
Comment (itikhono) [reply src/common/transformations/src/transformations/common_optimizations/augru_cell_fusion.cpp]: supported rzh format
As a result of the discussion, we decided to support rzh only

File: src/common/transformations/src/transformations/common_optimizations/augru_cell_fusion.cpp (hunk 12 above)
Comment (itikhono) [reply src/common/transformations/src/transformations/common_optimizations/augru_cell_fusion.cpp:30]: I prefer 0based indexes : )

**Review by mitruska — APPROVED** [review]

**Review by mitruska — COMMENTED** [review]
Update of onednn submodule should be removed from this PR as not related change.

File: src/common/transformations/src/transformations/common_optimizations/augru_cell_fusion.cpp (hunk 12 above)
Comment (itikhono) [reply src/common/transformations/src/transformations/common_optimizations/augru_cell_fusion.cpp:30]: is_**second**_dim_static exactly! 
thanks

Comment (itikhono) [reply src/common/transformations/src/transformations/common_optimizations/augru_cell_fusion.cpp]: fixed

**Review by ilyachur — APPROVED** [review]

File: src/common/transformations/src/transformations/common_optimizations/augru_cell_fusion.cpp [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+#include "itt.hpp"
+#include "ngraph_ops/augru_cell.hpp"
+#include "openvino/core/rt_info.hpp"
+#include "openvino/opsets/opset9.hpp"
+#include "openvino/pass/pattern/op/wrap_type.hpp"
+
+using namespace std;
+using namespace ov::opset9;
+using namespace ov::element;
+using namespace ov::pass::pattern;
```
[the remark this answers is not in the document]
Comment (itikhono) [reply src/common/transformations/src/transformations/common_optimizations/augru_cell_fusion.cpp:18]: but if we don't use it, the code looks too overloaded

**Pull request merged**